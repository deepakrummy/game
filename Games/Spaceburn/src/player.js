var whiten = require('./whiten');
var canvas = document.querySelector('#game');

window.player = {};

player.idle = new Image();
player.idle.src = 'images/astro.png';
player.idle.name = 'astro.png';
player.flying = new Image();
player.flying.src = 'images/astro-flying.png';
player.flying.name = 'astro-flying.png';
player.state = 'idle';

player.defaults = {
    money: 0,
    angle: 0,
    offsetY: 0,
    offsetX: 0,
    health: 100,
    fuel: 100,
    hit: false,
    propRange: 8.3,
    moneyMultiplier: 1
};

player.width = 52;
player.height = 60;
player.x = (canvas.width - player.width) / 2;
player.y = (canvas.height - player.height) / 2;
player.angle = 0;
player.totalMoney = player.money = 0;
player.colors = ['black', 'orange'];
player.defaults.equipped = 'explode';

player.offsetX = player.offsetY = 0;


// Half width, half height
var hW = player.width / 2;
var hH = player.height / 2;

var speed = 0; // The current speed
var dSpeed;
var dX = 0, dY = 0;

// YOU CAN CONFIGURE THESE! --------------------------
var acc = 7; // Acceleration
var lim = 10; // Speed limit
var turnSpeed = 2.2;
var grav = 0.03;
// NO MORE CONFIGURING! ------------------------------



player.reset = function() {
    dX = dY = speed = dSpeed = 0;
    var keys = Object.keys(player.defaults);
    for (var i = 0; i < keys.length; i++) {
        player[keys[i]] = player.defaults[keys[i]];
    }
    player.move();
};
player.gravity = function(elapsed) {
    dY -= grav;
};
player.move = function(elapsed, flying) {
    player.offsetX = dX;
    player.offsetY = -dY;
    dX *= 0.99;
    dY *= 0.99;

    if (!flying) {
        player.state = 'idle';
    }
};
player.up = function(elapsed) {
    player.fuel -= 0.2;
    player.state = 'flying';
    speed += acc;
    dSpeed = elapsed * speed;

    dX += Math.sin(player.angle) * dSpeed;
    dY += Math.cos(player.angle) * dSpeed;
    player.move(elapsed, true);
    if (speed > lim) {
        speed = lim;
    }
    else if (speed < -lim) {
        speed = -lim;
    }

};
player.right = function(elapsed) {
    player.angle += elapsed * turnSpeed * Math.PI;
};
player.left = function(elapsed) {
    player.angle -= elapsed * turnSpeed * Math.PI;
};
player.trigger = function() {
    // Enable the equippable item
    player.triggered = player.equipped;
};

player.addMoney = function(fo) {
    player.money += player.moneyMultiplier * (fo.width * fo.height) / 1000;
};


var ticks = 0;
player.draw = function(ctx) {
    // Player
    ctx.save();
    ctx.translate(player.x + hW, player.y + hH);
    ctx.rotate(player.angle);
    // player.hit is set in collisions.js
    // If the player's been hit, we want it to flash white to indicate that
    if (player.triggered === 'invincibility') {
        ctx.drawImage(whiten(player[player.state].name, 'green'), -hW, -hH);
    }
    else if (player.hit) {
        ctx.drawImage(whiten(player[player.state].name, 'pink'), -hW, -hH);
        ticks++;
        if (ticks >= 8) {
            player.hit = false;
            ticks = 0;
        }
    }
    else {
        ctx.drawImage(player[player.state], -hW, -hH);
    }
    ctx.restore();

};

player.reset();

module.exports = player;
