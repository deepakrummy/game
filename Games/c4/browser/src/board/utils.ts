import { Board } from './index'
/**
 * From Mozilla Developer Network
 * https://developer.mozilla.org/en-US/docs/Web/Events/resize
 */
export function onresize(): { add: Function } {
  const callbacks: Array<Function> = []
  let running: boolean = false

  // Fired on resize event.
  function resize() {
    if (!running) {
      running = true

      if (window.requestAnimationFrame)
        window.requestAnimationFrame(runCallbacks)
      else setTimeout(runCallbacks, 66)
    }
  }

  // Run the actual callbacks.
  function runCallbacks() {
    callbacks.forEach((callback: Function): void => {
      callback()
    })

    running = false
  }

  // Adds callback to loop.
  function addCallback(callback: Function): void {
    if (callback) callbacks.push(callback)
  }

  return {
    // Public method to add additional callback.
    add: (callback: Function) => {
      if (!callbacks.length) window.addEventListener('resize', resize)

      addCallback(callback)
    },
  }
}

export function drawCircle(
  context: CanvasRenderingContext2D,
  { x = 0, y = 0, r = 0, fillStyle = '', strokeStyle = '' }
) {
  context.save()
  context.fillStyle = fillStyle
  context.strokeStyle = strokeStyle
  context.beginPath()
  context.arc(x, y, r, 0, 2 * Math.PI, false)
  context.fill()
  context.restore()
}
/**
 * @see http://stackoverflow.com/a/11770000/917957
 * @param context Canvas 2D Context
 * @param board   current board
 */
export function drawMask(board: Board) {
  const context = board.context
  context.save()
  context.fillStyle = Board.MASK_COLOR
  context.beginPath()
  const doubleRadius = 2 * Board.PIECE_RADIUS
  const tripleRadius = 3 * Board.PIECE_RADIUS
  for (let y = 0; y < Board.ROWS; y++) {
    for (let x = 0; x < Board.COLUMNS; x++) {
      context.arc(
        tripleRadius * x + Board.MASK_X_BEGIN + doubleRadius,
        tripleRadius * y + Board.MASK_Y_BEGIN + doubleRadius,
        Board.PIECE_RADIUS,
        0,
        2 * Math.PI
      )
      context.rect(
        tripleRadius * x + Board.MASK_X_BEGIN + 2 * doubleRadius,
        tripleRadius * y + Board.MASK_Y_BEGIN,
        -2 * doubleRadius,
        2 * doubleRadius
      )
    }
  }
  context.fill()
  context.restore()
}

export function clearCanvas(board: Board) {
  board.context.clearRect(0, 0, Board.CANVAS_WIDTH, Board.CANVAS_HEIGHT)
}
