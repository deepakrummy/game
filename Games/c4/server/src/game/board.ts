import { BoardBase } from '@kenrick95/c4'

export class ServerBoard extends BoardBase {}
