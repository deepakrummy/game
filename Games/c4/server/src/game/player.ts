import { BoardPiece } from '@kenrick95/c4'
import { PlayerShadow } from '@kenrick95/c4'
import { PlayerId } from '../types'

export class ServerPlayer extends PlayerShadow {
  playerId: PlayerId
  constructor(boardPiece: BoardPiece, playerId: PlayerId) {
    super(boardPiece)
    this.playerId = playerId
  }
}
