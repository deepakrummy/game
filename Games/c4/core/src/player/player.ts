import { BoardBase, BoardPiece } from '../board'

export abstract class Player {
  boardPiece: BoardPiece
  /** @return {number} column number (0-index) */
  abstract getAction(board: BoardBase): Promise<number>
  constructor(boardPiece: BoardPiece) {
    this.boardPiece = boardPiece
  }
}
