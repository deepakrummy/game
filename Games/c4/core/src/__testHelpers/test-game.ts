import { GameBase } from '../game'
import { BoardBase } from '../board'
import { Player } from '../player'

export class TestGame extends GameBase {
  afterMoveResolve: null | ((action: number) => void) = null
  afterMovePromise: null | Promise<number> = null

  constructor(players: Array<Player>, board: BoardBase) {
    super(players, board)
    this.renewAfterMovePromise()
  }

  waitingForMove() {
    // no-op
  }
  beforeMoveApplied() {
    // no-op
  }
  afterMove(action: number) {
    if (this.afterMoveResolve) {
      this.afterMoveResolve(action)
    }
    this.renewAfterMovePromise()
  }

  renewAfterMovePromise() {
    this.afterMovePromise = new Promise(
      (resolve) => (this.afterMoveResolve = resolve)
    )
  }
}
