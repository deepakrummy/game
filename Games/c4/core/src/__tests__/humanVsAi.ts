import { TestGame } from '../__testHelpers/test-game'
import { BoardPiece, BoardBase } from '../board'
import { TestPlayer } from '../__testHelpers/test-player'
import { describe, test, expect } from 'vitest'
import { PlayerAi } from '../player'

describe('PlayerHuman vs PlayerAi', () => {
  const testPlayer = new TestPlayer(BoardPiece.PLAYER_1)
  const aiPlayer = new PlayerAi(BoardPiece.PLAYER_2)
  const players = [testPlayer, aiPlayer]
  test('Issue #3', async () => {
    const board = new BoardBase()
    const game = new TestGame(players, board)

    /*
    Next move is Player 2 (AI)
  
    Current board:

    0 0 0 0 0 0 0
    0 0 0 1 0 0 0
    0 0 0 1 0 1 0
    0 0 0 2 1 2 1
    0 0 0 1 2 2 2
    0 0 0 1 2 2 1

    Should choose column 2 to block immediate Player 1 from winning
    */
    board.map = [
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
      ],
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_1,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
      ],
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_1,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_1,
        BoardPiece.EMPTY,
      ],
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_1,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_1,
      ],
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_1,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_2,
      ],
      [
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.EMPTY,
        BoardPiece.PLAYER_1,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_2,
        BoardPiece.PLAYER_1,
      ],
    ]

    expect(board.map).toMatchSnapshot()
    game.currentPlayerId = 1
    game.start()
    const chosenAction = await game.afterMovePromise
    expect(chosenAction).toBe(2)
    expect(board.map).toMatchSnapshot()
  })
})
