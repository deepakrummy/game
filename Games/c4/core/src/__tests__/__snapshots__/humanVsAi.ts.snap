// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerHuman vs PlayerAi > Issue #3 1`] = `
[
  [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
    1,
    0,
    1,
    0,
  ],
  [
    0,
    0,
    0,
    2,
    1,
    2,
    1,
  ],
  [
    0,
    0,
    0,
    1,
    2,
    2,
    2,
  ],
  [
    0,
    0,
    0,
    1,
    2,
    2,
    1,
  ],
]
`;

exports[`PlayerHuman vs PlayerAi > Issue #3 2`] = `
[
  [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  [
    0,
    0,
    0,
    1,
    0,
    1,
    0,
  ],
  [
    0,
    0,
    0,
    2,
    1,
    2,
    1,
  ],
  [
    0,
    0,
    0,
    1,
    2,
    2,
    2,
  ],
  [
    0,
    0,
    2,
    1,
    2,
    2,
    1,
  ],
]
`;
